### PR TITLE
Refactoring measures

### PR DESCRIPTION
**Pull request by LFISHER7**
This refactors `measures.py` to make testing easier.

The main change is to produce a single measures file, which makes downstream actions cleaner and output checking easier.

### CODE REVIEW

File: interactive_templates/templates/v2/analysis/top_5.py [hunk 1]
```diff
@@ -82,9 +82,9 @@ def create_top_5_code_table(
         A table of the top `nrows` codes.
     """
 
-    # cast both code columns to str
-    df[code_column] = df[code_column].astype(int).astype(str)
-    code_df[code_column] = code_df[code_column].astype(int).astype(str)
+    # prevent scientific notation for codes by casting to float then int
```
Comment (ccunningham101) [inline interactive_templates/templates/v2/analysis/top_5.py:85]: Codelists with 17/18 codes are not that common, but I think you will run into the same thing we did here with codes being truncated. It happens as soon as the cohort-joiner step, so will the codes would already be truncated before reaching here. 

The strepa/scarlet solution was to make a local copy of the cohort joiner and [cast everything to strings ](https://github.com/opensafely/strepa_scarlet/blob/main/analysis/cohort_joiner.py#L12), which is not ideal from a memory usage perspective.

If you will always know the column name of the code column, you could provide it in a dictionary ([see docs](https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html))
Or you could use feather files

File: interactive_templates/templates/v2/analysis/report_utils.py [hunk 2]
```diff
@@ -1,10 +1,61 @@
...
+        rate_per: Value to calculate rate per
+        round: Bool indicating whether to round rate to 2dp.
+    """
+    if round_rate:
+        rate = round(df[value_col] / (df[population_col] / rate_per), 2)
+
+    else:
+        rate = df[value_col] / (df[population_col] / rate_per)
+
+    rate = rate.apply(round, ndigits=4)
```
Comment (ccunningham101) [inline interactive_templates/templates/v2/analysis/report_utils.py:38]: Is this redundant? Could supply ndigits=4 in the else statement?

File: interactive_templates/templates/v2/analysis/report_utils.py [hunk 3]
```diff
@@ -1,10 +1,61 @@
...
+        "M": "Male",
+    }
+
+    df = df.replace({"sex": sex_codes})
+    return df
+
+
+def generate_expectations_codes(codelist, incidence=0.5):
+    expectations = {str(x): (1 - incidence) / 10 for x in codelist[0:10]}
+    # expectations = {str(x): (1-incidence) / len(codelist) for x in codelist}
```
Comment (ccunningham101) [inline interactive_templates/templates/v2/analysis/report_utils.py:54]: Delete commented line

File: interactive_templates/templates/v2/analysis/measures.py [hunk 4]
```diff
@@ -2,7 +2,7 @@
 from pathlib import Path
 
 import pandas as pd
-from analysis.report_utils import get_date_input_file, match_input_files
+from analysis.report_utils import calculate_rate, get_date_input_file, match_input_files
 
 
 def round_column(df, col, decimals=-1):
```
Comment (ccunningham101) [inline interactive_templates/templates/v2/analysis/measures.py:8]: Could rename redact_and_round or round_and_redact, and the name "round" does not indicate that it also does redaction

File: interactive_templates/templates/v2/analysis/measures.py [hunk 5]
```diff
@@ -14,6 +14,117 @@ def round_column(df, col, decimals=-1):
...
+                "value",
+            ] = "[Redacted]"
+
+        result = pd.concat([result, group_df], ignore_index=True)
+
+    return result
+
+
 def parse_args():
     parser = argparse.ArgumentParser()
```
Comment (ccunningham101) [inline interactive_templates/templates/v2/analysis/measures.py:129]: args.measure is currently unused? Could be removed from and the call in the project.yaml

**Review by ccunningham101 — APPROVED** [review]
Some minor comments
And the outstanding issue of truncated codes, but that could be treated as a separate issue

File: interactive_templates/templates/v2/analysis/top_5.py (hunk 1 above)
Comment (LFISHER7) [reply interactive_templates/templates/v2/analysis/top_5.py:85]: I noticed in some of the outputs the code descriptions aren't being matched for all codes. I forgot with had the same issue in the strepa project, perhaps that's the issue here. I'll look at updating it

File: interactive_templates/templates/v2/analysis/report_utils.py (hunk 2 above)
Comment (LFISHER7) [reply interactive_templates/templates/v2/analysis/report_utils.py:38]: It is, i'll remove it

File: interactive_templates/templates/v2/analysis/report_utils.py (hunk 3 above)
Comment (LFISHER7) [reply interactive_templates/templates/v2/analysis/report_utils.py:54]: This doesn't actually do what it should for codelists with length <10. I'll update it

**Comment by LFISHER7** [comment]
Thanks @ccunningham101 , I'm going to address the problem with truncating codes separately.

**Pull request merged**